homeassistant/components/sensor/smappee.py

```diff
@@ -0,0 +1,177 @@
+"""
+Support for monitoring a Smappee energy sensor.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.smappee/
+"""
+import logging
+from datetime import datetime, timedelta
+
+from homeassistant.components.smappee import DATA_SMAPPEE
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+DEPENDENCIES = ['smappee']
+
+_LOGGER = logging.getLogger(__name__)
+
+SENSOR_PREFIX = 'Smappee'
+SENSOR_TYPES = {
+    'solar': ['Solar', 'mdi:white-balance-sunny', 'local', 'W'],
+    'alwaysOn': ['Always On', 'mdi:gauge', 'remote', 'W'],
+    'active_power': ['Active Power', 'mdi:power-plug', 'local', 'W'],
+    'current': ['Current', 'mdi:gauge', 'local', 'Amps'],
+    'voltage': ['Voltage', 'mdi:gauge', 'local', 'V'],
+    'active_cosfi': ['Power Factor', 'mdi:gauge', 'local', '%'],
+    'solar_today': ['Solar Today', 'mdi:white-balance-sunny', 'remote', 'kW'],
+    'power_today': ['Power Today', 'mdi:power-plug', 'remote', 'kW']
+}
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=30)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Smappee sensor."""
+    smappee = hass.data[DATA_SMAPPEE]
+
+    dev = []
+    if smappee.is_remote_active:
+        for sensor in SENSOR_TYPES:
+            if 'remote' in SENSOR_TYPES[sensor]:
+                for location_id in smappee.locations.keys():
+                    dev.append(SmappeeSensor(smappee, location_id, sensor))
+
+    if smappee.is_local_active:
+        for sensor in SENSOR_TYPES:
+            if 'local' in SENSOR_TYPES[sensor]:
+                if smappee.is_remote_active:
+                    for location_id in smappee.locations.keys():
+                        dev.append(SmappeeSensor(smappee, location_id, sensor))
+                else:
+                    dev.append(SmappeeSensor(smappee, None, sensor))
+    add_devices(dev)
+
+
+class SmappeeSensor(Entity):
+    """Implementation of a Smappee sensor."""
+
+    def __init__(self, smappee, location_id, sensor):
+        """Initialize the sensor."""
+        self._smappee = smappee
+        self._location_id = location_id
+        self._sensor = sensor
+        self.data = None
+        self._state = None
+        self._timestamp = None
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        if self._location_id:
+            location_name = self._smappee.locations[self._location_id]
+        else:
+            location_name = 'Local'
+
+        return "{} {} {}".format(SENSOR_PREFIX,
+                                 location_name,
+                                 SENSOR_TYPES[self._sensor][0])
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return SENSOR_TYPES[self._sensor][1]
```

Comment:
Can you cache this in the constructor so you don't have to do the same lookup on each update